tls/test_tls_limits.py

```diff
@@ -0,0 +1,109 @@
+"""
+Tests for Frang TLS limits.
+"""
+
+from framework import tester
+from helpers import remote, tf_cfg, util, dmesg
+
+
+__author__ = 'Tempesta Technologies, Inc.'
+__copyright__ = 'Copyright (C) 2020 Tempesta Technologies, Inc.'
+__license__ = 'GPL2'
+
+
+class TLSLimits(tester.TempestaTest):
+
+    clients = [
+        {
+            'id' : 'tls-perf',
+            'type' : 'external',
+            'binary' : 'tls-perf',
+            'cmd_args' : (
+                '-l 1 -t 1 -n 11  --tickets off ${server_ip} 443'
+                )
+        },
+    ]
+
+    backends = [
+        {
+            'id' : '0',
+            'type' : 'deproxy',
+            'port' : '8000',
+            'response' : 'static',
+            'response_content' :
+                'HTTP/1.1 200 OK\r\n'
+                'Content-Length: 0\r\n'
+                'Connection: keep-alive\r\n\r\n'
+        }
+    ]
+
+    tempesta = {
+        'config' : """
+            cache 0;
+            listen 443 proto=https;
+
+            tls_certificate ${tempesta_workdir}/tempesta.crt;
+            tls_certificate_key ${tempesta_workdir}/tempesta.key;
+
+            srv_group srv_grp1 {
+                server ${server_ip}:8000;
+            }
+            vhost tempesta-tech.com {
+                proxy_pass srv_grp1;
+            }
+            http_chain {
+                host == "tempesta-tech.com" -> tempesta-tech.com;
+                -> block;
+            }
+
+            frang_limits {
+                tls_connection_rate 10;
+            }
+        """
+    }
+
+    TLS_WARN="Warning: frang: new TLS connections rate exceeded for "
+
+    def test(self):
+        tls_perf = self.get_client('tls-perf')
+
+        self.start_all_servers()
+        self.start_tempesta()
+
+        klog = dmesg.DmesgFinder(ratelimited=False)
+
+        self.start_all_clients()
+        self.wait_while_busy(tls_perf)
+
+        self.assertEqual(klog.warn_count(self.TLS_WARN), 1,
+                         "Frang limits warning is not shown")
+
+
+class TLSLimitsBurst(TLSLimits):
+
+    tempesta = {
+        'config' : """
+            cache 0;
+            listen 443 proto=https;
+
+            tls_certificate ${tempesta_workdir}/tempesta.crt;
+            tls_certificate_key ${tempesta_workdir}/tempesta.key;
+
+            srv_group srv_grp1 {
+                server ${server_ip}:8000;
+            }
+            vhost tempesta-tech.com {
+                proxy_pass srv_grp1;
+            }
+            http_chain {
+                host == "tempesta-tech.com" -> tempesta-tech.com;
+                -> block;
+            }
+
+            frang_limits {
+                tls_connection_burst 10;
```

Comment:
The directive `tls_connection_rate` was added. 
Yes, `tls-perf`has no constant rate limit option, so it was impossible to test time frames with it. I have added extra tests using `deproxy` to simulate slower connections.  But test with `tls-perf` is still handy: it can use tickets to prove that resumed sessions are not limited by frang.